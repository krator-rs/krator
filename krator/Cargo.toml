[package]
name = "krator"
version = "0.3.0"
authors = [
    "Kevin Flansburg <kevin.flansburg@gmail.com>",
    "Taylor Thomas <taylor.thomas@microsoft.com>",
    "Matthew Fisher <matt.fisher@microsoft.com>",
    "Matt Butcher <matt.butcher@microsoft.com>",
]
edition = "2018"
license-file = "../LICENSE"
description = "A Kubernetes operator implementation in Rust"
repository = "https://github.com/krator-rs/krator"
readme = "README.md"
keywords = ["kubernetes", "controller", "operator", "state_machine"]

[badges]
maintenance = { status = "actively-developed" }

[features]
default = ["kube-native-tls"]
kube-native-tls = ["kube/native-tls", "kube-runtime/native-tls"]
rustls-tls = ["kube/rustls-tls", "kube-runtime/rustls-tls"]
derive = ["krator-derive"]
admission-webhook = ["warp", "json-patch", "serde_yaml"]
derive-admission-webhook = [
    "admission-webhook",
    "derive",
    "krator-derive/admission-webhook",
    "rcgen",
]

[dependencies]
async-trait = "0.1"
anyhow = "1.0"
tokio = { version = "1.0", features = ["fs", "macros", "signal"] }
tokio-stream = { version = "0.1", features = ['sync'] }
k8s-openapi = { version = "0.11", default-features = false, features = ["v1_20"] }
kube = { version = "0.55", default-features = false }
kube-runtime = { version = "0.55", default-features = false }
serde = { version = "1.0", features = ["derive"] }
serde_json = "1.0"
serde_yaml = { version = "0.8", optional = true }
futures = { version = "0.3", default-features = false }
krator-derive = { version = "0.2", path = "../krator-derive", optional = true }
warp = { version = "0.3", optional = true, features = ["tls"] }
json-patch = { version = "0.2", optional = true }
tracing = { version = "0.1", features = ['log'] }
tracing-futures = "0.2"
rcgen = { version = "0.8.9", features = ["x509-parser", "pem"], optional = true }

[dev-dependencies]
kube-derive = "0.55"
schemars = "0.8"
serde_yaml = "0.8"
chrono = "0.4"
rand = "0.8"
tracing-subscriber = "0.2"
tokio = { version = "1.0", features = ["fs", "macros", "signal", "rt-multi-thread"] }
opentelemetry-jaeger = "0.11"
tracing-opentelemetry = "0.11"
structopt = "0.3"

[[example]]
name = "moose"
required-features = ["derive"]
