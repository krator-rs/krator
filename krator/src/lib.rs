//! Exposes an API for creating state-machine-based Kubernetes Operators.

#![deny(missing_docs)]

mod manifest;
mod object;
mod operator;
mod runtime;
mod store;
pub mod util;

#[cfg(feature = "admission-webhook")]
pub mod admission;

pub mod state;

// TODO: Remove once webhooks are supported.
#[cfg(not(feature = "admission-webhook"))]
mod manager;
#[cfg(not(feature = "admission-webhook"))]
pub use manager::controller::ControllerBuilder;
#[cfg(not(feature = "admission-webhook"))]
pub use manager::Manager;

pub use manifest::Manifest;
pub use object::{ObjectState, ObjectStatus};
pub use operator::Operator;
pub use operator::Watchable;
pub use runtime::OperatorRuntime;
pub use state::{SharedState, State, Transition, TransitionTo};
pub use store::Store;

#[cfg(feature = "derive")]
#[allow(unused_imports)]
#[macro_use]
extern crate krator_derive;

#[cfg(feature = "derive")]
#[doc(hidden)]
pub use krator_derive::*;
