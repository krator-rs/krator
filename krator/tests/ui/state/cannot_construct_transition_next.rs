// Test that State<T> can only transition to State<T>
// edition:2018
extern crate async_trait;
extern crate krator;
extern crate k8s_openapi;

use krator::{Transition, state::StateHolder, ObjectState};
use k8s_openapi::apimachinery::pkg::apis::meta::v1::Status;
use k8s_openapi::api::core::v1::Pod;
use krator::state::test::Stub;

struct PodState;
struct ProviderState;

#[async_trait::async_trait]
impl ObjectState for PodState {
    type Manifest = Pod;
    type Status = Status;
    type SharedState = ProviderState;
    async fn async_drop(self, _provider_state: &mut ProviderState) { }
}

fn main() {
    // This fails because `state` is a private field. Use Transition::next classmethod instead.
    let _transition = Transition::<PodState>::Next(StateHolder {
        state: Box::new(Stub),
    });
}
