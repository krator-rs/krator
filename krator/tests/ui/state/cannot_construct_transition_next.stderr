error[E0451]: field `state` of struct `StateHolder` is private
  --> $DIR/cannot_construct_transition_next.rs:26:9
   |
26 |         state: Box::new(Stub),
   |         ^^^^^^^^^^^^^^^^^^^^^ private field

error: aborting due to previous error

For more information about this error, try `rustc --explain E0451`.
